src/sidebar/config/build-settings.js

```diff
@@ -29,129 +31,156 @@ function getAncestorFrame(levels, window_ = window) {
+function mergeSettings(configFromSidebar, configFromHost, rpcSettings) {
+  /** @type {SidebarSettings} */
   const sidebarSettings = {
-    ...appConfig,
-    ...hostPageConfig,
+    ...configFromSidebar,
+    ...configFromHost,
   };
+  if (rpcSettings) {
+    sidebarSettings.rpc = rpcSettings;
+  }
   sidebarSettings.apiUrl = getApiUrl(sidebarSettings);
   return sidebarSettings;
 }
 
 /**
- * Merge client configuration from h service with config fetched from
- * a parent frame asynchronously.
+ * Which groups to load and show in the sidebar may be dictated in
+ * ConfigFromHost configuration: the special value `$rpc:requestGroups`
+ * indicates that the list of groups should be requested asynchronously.
  *
- * Use this method to retrieve the config asynchronously from a parent
- * frame via RPC. See tests for more details.
+ * This function (maybe) mutates values in the provided `configFromHost`'s
+ * ServiceGroups: `$rpc:requestGroups` values are replaced with
+ * `Promise<string[]>`.
  *
- * @param {ConfigFromSidebar} appConfig - Settings rendered into `app.html` by the h service.
- * @param {Window} parentFrame - Frame to send call to.
- * @param {string} origin - Origin filter for `window.postMessage` call.
- * @return {Promise<SidebarSettings>} - The merged settings.
- */
-async function fetchConfigRpc(appConfig, parentFrame, origin) {
-  const remoteConfig = await postMessageJsonRpc.call(
-    parentFrame,
-    origin,
-    'requestConfig',
-    [],
-    3000
-  );
-  /**
-   * @param {string} method
-   * @param {any[]} args
-   */
-  const rpcCall = (method, args = [], timeout = 3000) =>
-    postMessageJsonRpc.call(parentFrame, origin, method, args, timeout);
-  const sidebarSettings = fetchConfigEmbed(appConfig, remoteConfig);
-  return fetchGroupsAsync(sidebarSettings, rpcCall);
-}
-
-/**
- * Potentially mutates the `groups` config object(s) to be a promise that
- * resolves right away if the `groups` value exists in the original
- * config, or returns a promise that resolves with a secondary RPC
- * call to `requestGroups` when the `groups` value is '$rpc:requestGroups'
- * If the `groups` value is missing or falsely then it won't be mutated.
+ * @see ServiceGroups
  *
- * This allows the app to start with an incomplete config and then lazily
- * fill in the `groups` value(s) later when its ready. This helps speed
- * up the loading process.
- *
- * @param {SidebarSettings} config - The configuration object to mutate. This should
- *  already have the `services` value
- * @param {function} rpcCall - RPC method
- *  (method, args, timeout) => Promise
- * @return {Promise<SidebarSettings>} - The mutated settings
+ * @param {ConfigFromHost} configFromHost
+ * @param {RPCSettings} rpcSettings
+ * @returns {ConfigFromHost}
  */
-async function fetchGroupsAsync(config, rpcCall) {
-  if (Array.isArray(config.services)) {
-    config.services.forEach((service, index) => {
+function fetchServiceGroups(configFromHost, rpcSettings) {
```

Comment:
In looking at this (and especially after typing it better), I couldn't see any reason that this function should be `async`.